docs/topics/multiplatform/multiplatform-compatibility-guide.md

```diff
@@ -490,3 +502,251 @@ projects.
+       val commonJvmMain by sourceSets.creating {
+           dependsOn(commonMain)
+           dependencies {
+               // shared dependencies
+           }
+       }
+       val jvmKtorMain by getting {
+           dependsOn(commonJvmMain)
+           dependencies {
+               // Ktor dependencies
+           }
+       }
+       val jvmOkioMain by getting {
+           dependsOn(commonJvmMain)
+           dependencies {
+               // OKIO dependencies
+           }
+       }
+   }
+}
+```
+
+You can see that it comes with a non-trivial configuration complexity:
+* You have to set-up Gradle Attributes both on the `:shared`-side and on each consumer side. Otherwise, Gradle won't be
+able to resolve dependencies on such projects, because without an additional information it's not clear whether the 
+consumer should receive Ktor-based implementation, or OKIO-based implementation
+* You have to set-up the `commonJvmMain` source set manually
+
+All this involves a handful of low-level Gradle and KGP abstractions and APIs. 
+
+**What's the best practice now?**
+
+The configuration gets pretty complicated because Ktor-based and OKIO-based implementations are
+**in the same Gradle project**. In a lot of cases, it is possible to extract those parts into separate Gradle projects.
+Here's the rough outline of this refactoring:
+
+1. Replace two duplicating targets from the original project to one target. If you had a shared source set between
+these targets, move its sources and configuration to the default source set of the newly created target:
+
+```kotlin
+// shared/build.gradle.kts
+kotlin {
+    jvm()
+    
+    sourceSets {
+        jvmMain {
+            // ... copy the configuration of jvmCommonMain here ...
+        }
+    }
+}
+```
+
+2. Add two new Gradle projects, usually by calling `include` in your `settings.gradle.kts`. For example:
+
+```kotlin
+include(":okio-impl")
+include(":ktor-impl")
+```
+
+3. Configure each new Gradle project
+   - Most likely, you don't need to apply `kotlin("multiplatform")` plugin, as those projects will compile only for one
+   target. In our case, we can apply `kotlin("jvm")`
+   - Move the content of original target-specific source sets to their respective projects, e.g. from `jvmKtorMain` to 
+   `ktor-impl/src`
+   - Copy the configuration of source sets (dependencies, compiler options, etc.)
+   - Add a dependency from a new Gradle project to the original one.
+
+```kotlin
+// ktor-impl/build.gradle.kts
+plugins {
+    kotlin("jvm")
+}
+
+dependencies {
+    project(":shared") // add dependency on the original project
+    // ... copy dependencies of jvmKtorMain here ...
+}
+
+kotlin {
+    compilerOptions {
+        // ... copy compiler options of jvmKtorMain here ...
+    }
+}
+```
+
+While this approach requires more initial setup work, it doesn't use any low-level entities of Gradle and KGP, making
+easier to work with the resulting build and maintain it.
```

Comment:
I think where this falls short is tests? Suppose you had a common test suite that tested both `ktor` and `okhttp` clients. The code in the tests is all the same, it just expects a HTTP client:

```kotlin
// commonMain
interface HttpClient { ... }
expect fun HttpClient(): HttpClient

// commonTest
@Test 
fun testClient() {
  assertThat(HttpClient().get("https://...").data).isEqualsTo("Hello World")
}
```

With multiple JVM targets, it was only a matter of defining 2 actuals:

```kotlin
// ktorMain
actual fun HttpClient() = KtorHttpClient()

// okHttpMain
actual fun HttpClient() = OkHttpClient()
```

Now with different Gradle projects, how would one do this? `expect`/`actual` do not cross Gradle projects boundaries. And duplicating the test code doesn't feel great. Any thoughts?

Reply:
That's true. The brutally honest description for this deprecation would've been something like: "if you've encountered this deprecation, and you don't see an easy way to migrate your project, then reach us and tell more about your case". 

But that would be a bit unfriendly, so we tried our best to provide some intrsuctions which applicable at least in some cases. It is expected that these instructions miss some cases, and we'll be happy to learn if someone encounters such cases in real projects

Reply:
We actually have this exact `expect`/`actual` scenario [in this integration test](https://github.com/apollographql/apollo-kotlin/blob/main/tests/engine/src/commonTest/kotlin/engines.kt#L4). I've turned this problem upside down but still not sure how to convert. Any guidance would be super welcome.

Reply:
Just took another look at it. In the current state, it's working with a single JVM target and 2 compilations but shows this warning:

```
w: Kotlin Source Set 'ktorTest' can't depend on 'commonTest' as they are from different Source Set Trees.
Please remove this dependency edge.
```

I'm not sure how to remove the common source set without breaking the `expect`/`actual`